Meme generator API

pdany1116 opened the pull request.

pdany1116 committed:
Initial structure of solution and test project

pdany1116 committed:
Create tests and update structure

pdany1116 committed:
Implement image downloader

pdany1116 committed:
Add memefier

pdany1116 committed:
Fix tests

pdany1116 committed:
Change redirect status code to perform GET request

pdany1116 committed:
Update folder structure

pdany1116 committed:
Refactor code

pdany1116 committed:
Improve meme body parser

pirvudoru commented on projects/02-meme-generator-api/solution/spec/lib/api_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      fit "responds with bad request" do
+        test_post
+
+        expect(last_response.status).to eq 400
```
It'd help if we'd also add assertions for the body to capture as much of the response expectations as possible.

pdany1116 replied:
This was resolved on week4 branch.
742061def77f4ca8118acb9f3d87118bf61a0dff

pirvudoru commented on projects/02-meme-generator-api/solution/spec/lib/api_spec.rb:
```diff
@@ -0,0 +1,114 @@
+        }.to_json
+      end
+
+      fit "responds with bad request" do
```
`it` not `fit`.

pdany1116 replied:
3d250421fea7fef10414a2c74d08469b239f4456

pirvudoru commented on projects/02-meme-generator-api/solution/spec/lib/api_spec.rb:
```diff
@@ -0,0 +1,114 @@
+    end
+
+    context "with invalid request body" do
+      # 'uri' is wrong spelled
```
This should be inside the context description.

pdany1116 replied:
2342511d7ad970a40dc3828235825c477a405eec

pirvudoru commented on projects/02-meme-generator-api/solution/run_api.rb:
```diff
@@ -0,0 +1,3 @@
+require "./lib/api"
+
+API::run
```
We use `::` when accessing constants and `.` when invoking methods.

pdany1116 replied:
042f7acb5a63ce637817ac2219238d54aa71299f

pirvudoru commented on projects/02-meme-generator-api/solution/spec/lib/api_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      it "responds with temporary redirect" do
+        test_post
+
+        expect(last_response.status).to eq 303
```
We should also ensure that the redirect `Location` is the expected one.

pdany1116 replied:
This was resolved on week4 branch.
742061def77f4ca8118acb9f3d87118bf61a0dff

pirvudoru commented on projects/02-meme-generator-api/solution/lib/api.rb:
```diff
@@ -0,0 +1,46 @@
+    body = MemeBodyParser.parse(request.body.read)
+    create_tmp_directory
+    image_path = TMP_PATH + body.file_name
+    Down.download(body.uri, destination: image_path)
```
These can be extracted & encapsulated in a service to properly abstractize.

pdany1116 replied:
05df683dad2da2ff97a8a5882e5de37b8fbf4ee1

pirvudoru commented on projects/02-meme-generator-api/solution/lib/api.rb:
```diff
@@ -0,0 +1,46 @@
+
+    redirect "/meme/#{body.file_name}", 303
+
+  rescue Down::Error, URI::Error
```
https://en.wikipedia.org/wiki/Leaky_abstraction

pdany1116 replied:
a47ceefcbc87bba22e720bd3a963d34db8e01114

pirvudoru commented on projects/02-meme-generator-api/solution/lib/api.rb:
```diff
@@ -0,0 +1,46 @@
+  rescue JSON::ParserError
+    [400, "Invalid request body syntax!"]
+  rescue
+    [500, "An unknown internal error occured!"]
```
If something bad happens, `500` will be thrown by default.

pdany1116 replied:
50df7077565fe47ef66f33504b4814a3600d387c

pirvudoru commented on projects/02-meme-generator-api/solution/lib/meme_body_parser.rb:
```diff
@@ -0,0 +1,16 @@
+    json_body = JSON.parse(body)
+    raise MemeBodyError("'meme' not found in body!") unless json_body["meme"]
+    raise MemeBodyError("'meme.uri' not found in body!") unless json_body["meme"]["uri"]
+    raise MemeBodyError("'meme.text' not found in body!") unless json_body["meme"]["text"]
```
This class is both a `parser` and a `validator`. 

https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#single-responsibility-principle

dorupirvu-ayla committed:
Add meme generator week 2 requirements (#7)

* Add requirements

* Add bcrypt example

* Add readme to sinatra example

pirvudoru committed:
Update status code in meme generator requirements

Reason is that a 307 status code makes the client retry the exact request (POST with Body) on the redirect location.

lucianghinda committed:
Merge pull request #9 from Agilefreaks/update-staus-code

Update status code in meme generator requirements

pdany1116 committed:
Remove forgotten fit

pdany1116 committed:
Fix meme body parser and add tests

pdany1116 committed:
Fix API run method call

pdany1116 committed:
Create tmp directory before application start

pdany1116 committed:
Add image downloader

pdany1116 committed:
Update memefier to use tmp path

pdany1116 committed:
Map json parser error

pdany1116 committed:
Remove comment

pdany1116 committed:
Refactor api

pdany1116 committed:
500 is thrown by default

pdany1116 committed:
Merge branch 'Agilefreaks:main' into week3
